### ISSUE TITLE
Failed data migration repopulate-card-read-permissions while upgrading 0.28.1 -> 0.28.5

### ISSUE THREAD

**Issue by snarfed** [metabase/metabase#7333]
hi all! i recently upgraded from 0.28.1 to 0.28.5. our staging instance upgraded fine, but prod died during the `repopulate-card-read-permissions` data migration. i'm guessing we have an unhappy card in prod. here's my guess at the relevant snippets from the log:

```
04-11 23:54:32 INFO metabase.db :: Checking if Database has unrun migrations...
04-11 23:54:41 INFO metabase.db :: Database Migrations Current ...  ✅
com.mchange.v2.cfg.DelayedLogItem [ level -> FINE, text -> "The configuration file for resource identifier 'hocon:/reference,/application,/c3p0,/' could not be found. Skipping.", exception -> null]
04-11 23:54:41 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
04-11 23:54:42 INFO db.migrations :: Running data migration 'repopulate-card-read-permissions'...
clojure.lang.ExceptionInfo: Output of field does not match schema: 

	   (named (not (matches-some-precondition? nil)) "AnyField: field, ag field reference, expression, expression reference, or field literal.")  
```

i rolled back to 0.28.1 and it started up fine. full log below. this is running the bare jar on ubuntu 14.04 on AWS EC2. database is postgres 9.3 on RDS. not urgent. thanks in advance!

```
04-11 23:53:34 INFO metabase.util :: Loading Metabase...
04-11 23:53:57 INFO util.encryption :: DB details encryption is DISABLED for this Metabase instance. 🔓 
See http://www.metabase.com/docs/latest/operations-guide/start.html#encrypting-your-database-connection-details-at-rest for more information.
04-11 23:54:24 INFO metabase.core :: Starting Metabase in STANDALONE mode
****************************************************************************************************************************************************
*
* DEPRECATION WARNING: You are currently running JDK '1.7.0_95'. Support for Java 7 has been deprecated and will be dropped from a future release.
* See the operation guide for more information: https://metabase.com/docs/latest/operations-guide/start.html.
*
****************************************************************************************************************************************************
04-11 23:54:24 WARN metabase.core :: [DEPRECATION WARNING: You are currently running JDK '1.7.0_95'. Support for Java 7 has been deprecated and will be dropped from a future release. See the operation guide for more information: https://metabase.com/docs/latest/operations-guide/start.html.]
04-11 23:54:24 INFO metabase.core :: Launching Embedded Jetty Webserver with config:
 {:port 3000}

04-11 23:54:25 INFO metabase.core :: Starting Metabase version v0.28.5 (fa5cb8a release-0.28.5) ...
04-11 23:54:25 INFO metabase.core :: System timezone is 'Etc/UTC' ...
04-11 23:54:31 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
04-11 23:54:31 INFO metabase.db :: Verifying postgres Database Connection ...
04-11 23:54:31 INFO metabase.db :: Verify Database Connection ...  ✅
04-11 23:54:31 INFO metabase.db :: Running Database Migrations...
04-11 23:54:31 INFO metabase.db :: Setting up Liquibase...
04-11 23:54:32 INFO metabase.db :: Liquibase is ready.
04-11 23:54:32 INFO metabase.db :: Checking if Database has unrun migrations...
04-11 23:54:41 INFO metabase.db :: Database Migrations Current ...  ✅
com.mchange.v2.cfg.DelayedLogItem [ level -> FINE, text -> "The configuration file for resource identifier 'hocon:/reference,/application,/c3p0,/' could not be found. Skipping.", exception -> null]
04-11 23:54:41 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
04-11 23:54:42 INFO db.migrations :: Running data migration 'repopulate-card-read-permissions'...
clojure.lang.ExceptionInfo: Output of field does not match schema: 

	   (named (not (matches-some-precondition? nil)) "AnyField: field, ag field reference, expression, expression reference, or field literal.")  

 {:type :schema.core/error, :schema (named (cond-pre (record metabase.query_processor.interface.Field {:description (maybe (constrained Str "Non-blank string")), :base-type (pred "Valid field type"), :parent Any, :table-id (constrained Int "Integer greater than zero"), :special-type (maybe (pred "Valid field type")), :remapped-from (maybe Str), :remapped-to (maybe Str), :field-name (constrained Str "Non-blank string"), :database-type (constrained Str "Non-blank string"), :dimensions (maybe (cond-pre (record metabase.query_processor.interface.Dimensions {:dimension-id (constrained Int "Integer greater than zero"), :field-id (constrained Int "Integer greater than zero"), :dimension-name (constrained Str "Non-blank string"), :human-readable-field-id (maybe (constrained Int "Integer greater than zero")), :dimension-type (enum :external :internal), :created-at Inst, :updated-at Inst}) {} [])), :field-display-name (constrained Str "Non-blank string"), :values (maybe (cond-pre (record metabase.query_processor.interface.FieldValues {:field-value-id (constrained Int "Integer greater than zero"), :field-id (constrained Int "Integer greater than zero"), :values (maybe (cond-pre [Any] {} [])), :human-readable-values (maybe (cond-pre [Any] {} [])), :created-at Inst, :updated-at Inst}) {} [])), :parent-id (maybe (constrained Int "Integer greater than zero")), :visibility-type (enum :retired :sensitive :normal :hidden :details-only), :position (maybe (constrained Int "Integer greater than zero")), :field-id (constrained Int "Integer greater than zero"), :table-name (maybe (constrained Str "Non-blank string")), :fk-field-id (maybe Int), :fingerprint (maybe {(optional-key :global) {(optional-key :distinct-count) Int}, (optional-key :type) (constrained {(optional-key :type/Number) {(optional-key :min) Num, (optional-key :max) Num, (optional-key :avg) Num}, (optional-key :type/Text) {(optional-key :percent-json) (constrained Num "Valid percentage between (inclusive) 0 and 1."), (optional-key :percent-url) (constrained Num "Valid percentage between (inclusive) 0 and 1."), (optional-key :percent-email) (constrained Num "Valid percentage between (inclusive) 0 and 1."), (optional-key :average-length) (constrained java.lang.Double "Valid number greater than or equal to zero")}} "Type-specific fingerprint with exactly one key"), (optional-key :experimental) {Keyword Any}}), :schema-name (maybe (constrained Str "Non-blank string"))}) (record metabase.query_processor.interface.FieldPlaceholder {:field-id (constrained Int "Integer greater than zero"), :fk-field-id (maybe (constrained (constrained Int "Integer greater than zero") "foreign-keys is not supported by this driver.")), :datetime-unit (maybe (named (enum :quarter :day :hour :week :default :day-of-week :hour-of-day :month :month-of-year :day-of-month :year :day-of-year :week-of-year :quarter-of-year :minute-of-hour :minute) "Valid datetime unit for a field")), :remapped-from (maybe Str), :remapped-to (maybe Str), :field-display-name (maybe Str), :binning-strategy (maybe (enum :default :bin-width :num-bins)), :binning-param (maybe Num)}) (record metabase.query_processor.interface.DateTimeField {:field (cond-pre (record metabase.query_processor.interface.Field {:description (maybe (constrained Str "Non-blank string")), :base-type (pred "Valid field type"), :parent Any, :table-id (constrained Int "Integer greater than zero"), :special-type (maybe (pred "Valid field type")), :remapped-from (maybe Str), :remapped-to (maybe Str), :field-name (constrained Str "Non-blank string"), :database-type (constrained Str "Non-blank string"), :dimensions (maybe (cond-pre (record metabase.query_processor.interface.Dimensions {:dimension-id (constrained Int "Integer greater than zero"), :field-id (constrained Int "Integer greater than zero"), :dimension-name (constrained Str "Non-blank string"), :human-readable-field-id (maybe (constrained Int "Integer greater than zero")), :dimension-type (enum :external :internal), :created-at Inst, :updated-at Inst}) {} [])), :field-display-name (constrained Str "Non-blank string"), :values (maybe (cond-pre (record metabase.query_processor.interface.FieldValues {:field-value-id (constrained Int "Integer greater than zero"), :field-id (constrained Int "Integer greater than zero"), :values (maybe (cond-pre [Any] {} [])), :human-readable-values (maybe (cond-pre [Any] {} [])), :created-at Inst, :updated-at Inst}) {} [])), :parent-id (maybe (constrained Int "Integer greater than zero")), :visibility-type (enum :retired :sensitive :normal :hidden :details-only), :position (maybe (constrained Int "Integer greater than zero")), :field-id (constrained Int "Integer greater than zero"), :table-name (maybe (constrained Str "Non-blank string")), :fk-field-id (maybe Int), :fingerprint (maybe {(optional-key :global) {(optional-key :distinct-count) Int}, (optional-key :type) (constrained {(optional-key :type/Number) {(optional-key :min) Num, (optional-key :max) Num, (optional-key :avg) Num}, (optional-key :type/Text) {(optional-key :percent-json) (constrained Num "Valid percentage between (inclusive) 0 and 1."), (optional-key :percent-url) (constrained Num "Valid percentage between (inclusive) 0 and 1."), (optional-key :percent-email) (constrained Num "Valid percentage between (inclusive) 0 and 1."), (optional-key :average-length) (constrained java.lang.Double "Valid number greater than or equal to zero")}} "Type-specific fingerprint with exactly one key"), (optional-key :experimental) {Keyword Any}}), :schema-name (maybe (constrained Str "Non-blank string"))}) (record metabase.query_processor.interface.FieldLiteral {:field-name (constrained Str "Non-blank string"), :base-type (pred "Valid field type")})), :unit (named (enum :quarter :day :hour :week :default :day-of-week :hour-of-day :month :month-of-year :day-of-month :year :day-of-year :week-of-year :quarter-of-year :minute-of-hour :minute) "Valid datetime unit for a field")}) (record metabase.query_processor.interface.FieldLiteral {:field-name (constrained Str "Non-blank string"), :base-type (pred "Valid field type")}) (record metabase.query_processor.interface.AgFieldRef {:index Int}) (record metabase.query_processor.interface.Expression {:operator (named (enum :* :- :/ :+) "Valid expression operator"), :args [(cond-pre (recursive (var metabase.query-processor.interface/AnyValueLiteral)) (recursive (var metabase.query-processor.interface/AnyField)) (recursive (var metabase.query-processor.interface/Aggregation)))], :custom-name (maybe (constrained Str "Non-blank string"))}) (record metabase.query_processor.interface.ExpressionRef {:expression-name (constrained Str "Non-blank string")})) "AnyField: field, ag field reference, expression, expression reference, or field literal."), :value nil, :error (named (not (matches-some-precondition? nil)) "AnyField: field, ag field reference, expression, expression reference, or field literal.")}
	at metabase.query_processor.middleware.expand$fn__28680$field__28685.invoke(expand.clj:43)
	at metabase.query_processor.middleware.expand$field_or_expression.invokeStatic(expand.clj:159)
	at metabase.query_processor.middleware.expand$field_or_expression.invoke(expand.clj:149)
	at metabase.query_processor.middleware.expand$fn__28955$ag_with_field__28960$fn__28961.invoke(expand.clj:162)
	at metabase.query_processor.middleware.expand$fn__28955$ag_with_field__28960.invoke(expand.clj:161)
	at clojure.core$partial$fn__4759.invoke(core.clj:2515)
	at clojure.lang.Var.invoke(Var.java:379)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.core$apply.invokeStatic(core.clj:646)
	at clojure.core$apply.invoke(core.clj:641)
	at metabase.query_processor.middleware.expand$fn__29696$expand_ql_sexpr__29701$fn__29705.invoke(expand.clj:553)
	at metabase.query_processor.middleware.expand$fn__29696$expand_ql_sexpr__29701.invoke(expand.clj:545)
	at metabase.query_processor.middleware.expand$walk_expand_ql_sexprs.invokeStatic(expand.clj:561)
	at metabase.query_processor.middleware.expand$walk_expand_ql_sexprs.invoke(expand.clj:556)
	at metabase.query_processor.middleware.expand$fn__29741$expand_inner__29746$fn__29747.invoke(expand.clj:569)
	at metabase.query_processor.middleware.expand$fn__29741$expand_inner__29746.invoke(expand.clj:565)
	at clojure.core$update.invokeStatic(core.clj:5960)
	at clojure.core$update.invoke(core.clj:5952)
	at metabase.query_processor.middleware.expand$expand.invokeStatic(expand.clj:595)
	at metabase.query_processor.middleware.expand$expand.invoke(expand.clj:580)
	at metabase.query_processor.middleware.expand$expand_middleware$fn__29767.invoke(expand.clj:602)
	at clojure.core$comp$fn__4727.invoke(core.clj:2460)
	at clojure.core$comp$fn__4727.invoke(core.clj:2460)
	at clojure.core$comp$fn__4727.invoke(core.clj:2460)
	at metabase.models.card$mbql_permissions_path_set.invokeStatic(card.clj:174)
	at metabase.models.card$mbql_permissions_path_set.doInvoke(card.clj:137)
	at clojure.lang.RestFn.invoke(RestFn.java:464)
	at metabase.models.card$query_perms_set.invokeStatic(card.clj:200)
	at metabase.models.card$query_perms_set.doInvoke(card.clj:191)
	at clojure.lang.RestFn.invoke(RestFn.java:464)
	at metabase.models.card$maybe_update_read_permissions.invokeStatic(card.clj:315)
	at metabase.models.card$maybe_update_read_permissions.invoke(card.clj:302)
	at metabase.models.card$pre_update.invokeStatic(card.clj:351)
	at metabase.models.card$pre_update.invoke(card.clj:349)
	at clojure.core$comp$fn__4727.invoke(core.clj:2460)
	at toucan.models$fn__3985$G__3904__3990.invoke(models.clj:178)
	at toucan.models$do_pre_update.invokeStatic(models.clj:330)
	at toucan.models$do_pre_update.invoke(models.clj:325)
	at toucan.db$update_BANG_.invokeStatic(db.clj:479)
	at toucan.db$update_BANG_.invoke(db.clj:462)
	at metabase.db.migrations$repopulate_card_read_permissions$fn__32982.invoke(migrations.clj:396)
	at clojure.core$run_BANG_$fn__7276.invoke(core.clj:7393)
	at clojure.core$completing$fn__6948.invoke(core.clj:6585)
	at clojure.core$map$fn__4781$fn__4782.invoke(core.clj:2633)
	at clojure.java.jdbc$init_reduce_rs.invokeStatic(jdbc.clj:1120)
	at clojure.java.jdbc$init_reduce_rs.invoke(jdbc.clj:1112)
	at clojure.java.jdbc$reducible_result_set_STAR_$reify__1874.reduce(jdbc.clj:1142)
	at clojure.core$reduce.invokeStatic(core.clj:6544)
	at clojure.core$reduce.invoke(core.clj:6527)
	at clojure.java.jdbc$query_reducer$fn__1880$fn__1883.invoke(jdbc.clj:1177)
	at clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:994)
	at clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:987)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1016)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:996)
	at clojure.java.jdbc$reducible_query$reify__1895.reduce(jdbc.clj:1272)
	at clojure.core$transduce.invokeStatic(core.clj:6600)
	at clojure.core.Eduction.reduce(core.clj:7360)
	at clojure.core$reduce.invokeStatic(core.clj:6544)
	at clojure.core$run_BANG_.invokeStatic(core.clj:7388)
	at clojure.core$run_BANG_.invoke(core.clj:7388)
	at metabase.db.migrations$repopulate_card_read_permissions.invokeStatic(migrations.clj:394)
	at metabase.db.migrations$repopulate_card_read_permissions.invoke(migrations.clj:393)
	at metabase.db.migrations$run_migration_if_needed_BANG_.invokeStatic(migrations.clj:50)
	at metabase.db.migrations$run_migration_if_needed_BANG_.invoke(migrations.clj:41)
	at metabase.db.migrations$run_all_BANG_.invokeStatic(migrations.clj:70)
	at metabase.db.migrations$run_all_BANG_.invoke(migrations.clj:64)
	at clojure.lang.Var.invoke(Var.java:375)
	at metabase.db$run_data_migrations_BANG_.invokeStatic(db.clj:390)
	at metabase.db$run_data_migrations_BANG_.invoke(db.clj:385)
	at metabase.db$setup_db_BANG_.invokeStatic(db.clj:401)
	at metabase.db$setup_db_BANG_.doInvoke(db.clj:392)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at metabase.core$init_BANG_.invokeStatic(core.clj:154)
	at metabase.core$init_BANG_.invoke(core.clj:133)
	at metabase.core$start_normally.invokeStatic(core.clj:258)
	at metabase.core$start_normally.invoke(core.clj:251)
	at metabase.core$_main.invokeStatic(core.clj:279)
	at metabase.core$_main.doInvoke(core.clj:274)
	at clojure.lang.RestFn.invoke(RestFn.java:397)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at metabase.core.main(Unknown Source)
04-11 23:54:42 ERROR metabase.core :: Metabase Initialization FAILED:  Output of field does not match schema: 

	   (named (not (matches-some-precondition? nil)) "AnyField: field, ag field reference, expression, expression reference, or field literal.")  


04-11 23:54:42 INFO metabase.core :: Metabase Shutting Down ...
04-11 23:54:42 INFO metabase.core :: Metabase Shutdown COMPLETE
```

**Comment by snarfed** [comment]
looking at existing issues, #7303 and #7304 seem to be about the same data migration, but different bugs.

**Comment by camsaul** [comment]
#7313 should fix your issue, please give it a try

**Comment by camsaul** [comment]
This was fixed by #7313
